src/Repository/UserRepository.php

```diff
@@ -170,17 +153,54 @@ public function countByCriteria(FindUserCriteria $criteria = null): int
                 ->setParameter('lineItemSlugs', $criteria->getLineItemSlugCriterion());
         }
 
-        if ($criteria->hasEuclideanDivisionCriterion()) {
-            $queryBuilder
-                ->where('MOD(u.id, :modulo) = :remainder')
-                ->setParameter('modulo', $criteria->getEuclideanDivisionCriterion()->getModulo())
-                ->setParameter('remainder', $criteria->getEuclideanDivisionCriterion()->getRemainder());
-        }
-
         $result = $queryBuilder
             ->getQuery()
             ->getOneOrNullResult();
 
         return null === $result ? 0 : (int)$result['number_of_users'];
     }
+
+    /**
+     * @throws ORMException
+     * @throws MappingException
+     */
+    public function insertMultipleNatively(UserDtoCollection $users): void
+    {
+        $queryParts = [];
+        foreach ($users as $user) {
+            $queryParts[] = sprintf(
+                "('%s', '%s', '%s', '[]', '%s')",
+                $user->getId(),
+                $user->getUsername(),
+                $user->getPassword(),
+                $user->getGroupId()
+            );
+        }
+
+        $query = sprintf(
+            'INSERT INTO users (id, username, password, roles, group_id) VALUES %s',
+            implode(',', $queryParts)
+        );
+
+        $this->_em->createNativeQuery($query, new ResultSetMapping())->execute();
+        $this->_em->clear();
+    }
+
+    /**
+     * @param string[] $usernames
+     */
+    public function findUsernames(array $usernames): array
```

Comment:
Suggestion: use findByUsernames instead.

Reply:
I have a question. What is the reason to do not return the collection of users instead of these raw values?

Reply:
To have better performance. This query is used in cache warmup, it would be too much to hydrate all the user objects, there is no need for it.